cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ResourceDistributionGoal.java

```diff
@@ -321,22 +389,180 @@ protected boolean rebalanceByMovingLoadIn(Broker broker,
+                .collect(Collectors.toSet()));
+
+    return candidateReplicasToSwapWith;
+  }
+
+  private boolean rebalanceBySwappingLoadOut(Broker broker,
                                              ClusterModel clusterModel,
                                              Set<Goal> optimizedGoals,
-                                             ActionType actionType,
-                                             Set<String> excludedTopics)
-      throws AnalysisInputException, ModelInputException {
-    // Get th eligible brokers.
-    SortedSet<Broker> candidateBrokers = new TreeSet<>((b1, b2) -> {
-        int result = Double.compare(utilizationPercentage(b1), utilizationPercentage(b2));
-        return result != 0 ? result : Integer.compare(b1.id(), b2.id());
+                                             Set<String> excludedTopics) {
+    if (!broker.isAlive()) {
+      return true;
+    }
+    // Get the replicas to rebalance.
+    SortedSet<Replica> sourceReplicas = new TreeSet<>((r1, r2) -> {
+      int result = Double.compare(r2.load().expectedUtilizationFor(resource()), r1.load().expectedUtilizationFor(resource()));
+      return result == 0 ? r1.topicPartition().toString().compareTo(r2.topicPartition().toString()) : result;
     });
+
+    sourceReplicas.addAll(resource() == Resource.NW_OUT ? broker.leaderReplicas() : broker.replicas());
+
+    // Sort the replicas initially to avoid sorting it every time.
+    PriorityQueue<CandidateBroker> candidateBrokerPQ = new PriorityQueue<>();
+    for (Broker candidate : clusterModel.healthyBrokersUnderThreshold(resource(), balanceUpperThreshold(clusterModel))
+                                        .stream().filter(b -> !b.replicas().isEmpty()).collect(Collectors.toSet())) {
+      // Get candidate replicas on candidate broker to try swapping with -- sorted in the order of trial.
+      double maxSourceReplicaLoad = sourceReplicas.first().load().expectedUtilizationFor(resource());
+      SortedSet<Replica> candidateReplicasToSwapWith = replicasToSwapWith(candidate, excludedTopics, maxSourceReplicaLoad, false);
+
+      CandidateBroker candidateBroker = new CandidateBroker(candidate, candidateReplicasToSwapWith, true);
+      candidateBrokerPQ.add(candidateBroker);
+    }
+
+    while (!candidateBrokerPQ.isEmpty()) {
+      CandidateBroker cb = candidateBrokerPQ.poll();
+      SortedSet<Replica> candidateReplicasToSwapWith = cb.replicas();
+
+      Replica swappedInReplica = null;
+      Replica swappedOutReplica = null;
+      for (Replica sourceReplica : sourceReplicas) {
+        if (excludedTopics.contains(sourceReplica.topicPartition().topic())) {
+          continue;
+        }
+
+        // Try swapping the source with the candidate replicas. Get the swapped in replica if successful, null otherwise.
+        Replica swappedIn = maybeApplySwapAction(clusterModel, sourceReplica, candidateReplicasToSwapWith, optimizedGoals);
```

Comment:
It seems we may "over-swap" with this logic? i.e. say the ideal utilization is 20, lower limit is 10, upper limit is 30 and current broker is 40, we may directly go to 10 instead of going 20?

Reply:
This goal aims to keep the load on the broker between 10 and 30; hence, there is no logic that explicitly prefers going directly to 20 rather than 10 -- i.e. 20 is not considered to be more balanced compared to 10.

Also note that going to 10 (or 20) does not necessarily mean that the broker will preserve this as its final load at the end of the rebalance. During this balancing process, the balancing actions may change the load on a balanced broker _as long as they are not making the balance worse_. This means that a broker whose load was 10 at the end of optimizing the resource distribution goal, may have a load of 20 at the end of the optimization of succeeding goals. Also, note critically that the load on the broker can never go from 20 to 10 once it is balanced. Hence, it would arguably be even more preferable to have the initial goal balance not to set the load to 20 directly due to the additional constraints that it would bring to the succeeding goals -- i.e. cannot make the balance worse.